Migrates plugin documentation into DocC (#8766)
apple/swift-package-manager: Sources/PackageManagerDocs/Documentation.docc/Plugins/WritingBuildToolPlugin.md

# Writing a build tool plugin

@Metadata {
    @Available("Swift", introduced: "5.6")
}

Create a build tool to process or generate files.

## Overview

The first step when writing a package plugin is to decide what kind of plugin you need.
Implement a build tool plugin to generate source files that should be part of a build, or to perform other actions at the start of every build.
Build tool plugins are invoked before a package is built in order to construct command invocations to run as part of the build.

A build tool plugin can provide two kinds of commands:

- term prebuild commands: commands that package manager runs before the build starts. Prebuild commands can generate an arbitrary number of output files with names that can't be predicted before running the command.
- term build commands: commands that package manager incorporates into the build system's dependency graph and runs at the appropriate time during the build based on the existence and timestamps of their predefined inputs and outputs.

> > Note: If your goal is to provide an actions that you can perform at any time and are not associated with a build, implement a command plugin.

Reviewer: typo:
```suggestion
> Note: If your goal is to provide an action that you can perform at any time and is not associated with a build, implement a command plugin.
```

> See <doc:WritingCommandPlugin> for details about creating a command plugin.

With both prebuild and build commands, it is important to note that the build tool plugin doesn't do the work, rather it constructs the commands that the build runs later, and it is those commands that perform the work.
The plugin can be quite small, and is often concerned with forming the command line for the build command that does the actual work.

A build tool plugin is available to the package that defines it, and if there is a corresponding plugin product, it is also available to any other package that has a direct dependency on the package.

### Build commands

Prefer to create a build command over a prebuild command when the paths of all of the inputs and outputs are known before the command runs.

> Build tool command are more efficient because they provide the build system the information needed to efficiently decide when the build should invoke them.

Reviewer: ```suggestion
Build tool commands are more efficient because they provide the build system the information needed to efficiently decide when the build should invoke them.
```

The following example illustrates defining a build tool named "MyBuildToolPlugin" that depends on the product `SomeTool`, and can be used from other packages:

```swift
import PackageDescription

let package = Package(
    name: "MyPluginPackage",
    products: [
        .plugin(
            name: "MyBuildToolPlugin",
            targets: [
                "MyBuildToolPlugin"
            ]
        )
    ],
    dependencies: [
        .package(
            url: "https://github.com/example/sometool",
            from: "0.1.0"
        )
    ],
    targets: [
        .plugin(
            name: "MyBuildToolPlugin",
            capability: .buildTool(),
            dependencies: [
                .product(name: "SomeTool", package: "sometool"),
            ]
        )
    ]
)
```

The [plugin target](https://developer.apple.com/documentation/packagedescription/target/plugin(name:capability:dependencies:path:exclude:sources:)) declares the name and capability of the plugin, along with its dependencies.
The capability of `.buildTool()` is the declaration that this defines a build tool plugin.
The capability also indicates the entry point the plugin is expected to implement.

> When you declare a `plugin` product, that makes the plugin visible to other packages that have a dependencu on the package.

Reviewer: ```suggestion
When you declare a `plugin` product, that makes the plugin visible to other packages that have a dependency on the package.
```

The intent is to eventually support any type of file that you could include as a source file in the target, and to let the plugin provide greater control over the downstream processing of generated files.

### Diagnostics

Plugin entry points are marked `throws`, and any errors thrown from the entry point cause the build system to mark the plugin invocation as failed.
Package manager presents the thrown error to the user, which should include a clear description of what went wrong.

Additionally, plugins can use the `Diagnostics` API in PackagePlugin to emit warnings and errors.
These optionally include references to file paths and line numbers in those files.

### Debugging and Testing

Package manager doesn't currently have any specific support for debugging and testing plugins.
Many plugins act as adapters that construct command lines for invoking the tools that do the real work.
In the cases in which there is non-trivial code in a plugin, a good approach is to factor out that code into separate source files that can be included in unit tests using symbolic links with relative paths.

### Xcode Extensions to the PackagePlugin API

> When you invoke a plugin in Apple’s Xcode IDE, the plugins has access to a library module provided by Xcode called *XcodeProjectPlugin*. 

Reviewer: ```suggestion
When you invoke a plugin in Apple’s Xcode IDE, the plugins have access to a library module provided by Xcode called *XcodeProjectPlugin*. 
```
